proto/streamlit/proto/MetricsEvent.proto

```diff
@@ -0,0 +1,85 @@
+/**!
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto3";
+
+option java_package = "com.snowflake.apps.streamlit";
+option java_outer_classname = "MetricsEventProto";
+
+import "streamlit/proto/PageProfile.proto";
+
+// Metrics events:
+message MetricsEvent {
+
+  // Common Event Fields:
+  string event = 1;
+  string anonymous_id = 2;
+  string machine_id_v3 = 3;
+  string report_hash = 4;
+  bool dev = 5;
+  string source = 6;
+  string streamlit_version = 7;
+  bool is_hello = 8;
+
+    // Host tracking fields:
+  string hosted_at = 9;
+  string owner = 10;
+  string repo = 11;
+  string branch = 12;
+  string main_module = 13;
+  string creator_id = 14;
+
+    // Context fields:
+  string context_page_url = 15;
+  string context_page_title = 16;
+  string context_page_path = 17;
+  string context_page_referrer = 18;
+  string context_page_search = 19;
+  string context_locale = 20;
+  string context_user_agent = 21;
+
+
+  // Menu Click Event field:
+  string label = 22;
+
+
+  // Page Profile Event fields:
```

Comment:
Do we need to keep this a flat structure? If nesting is fine we could potentially use an `oneof` here to have a better separation between the different event types, e.g.:

```proto
  oneof metadata {
    PageProfile page_profile = 1;
    MenuClick menu_click = 2;
```

In that case, we could potentially also reuse the existing page profile message by moving all the frontend added fields into it as well.

Reply:
I had same thought, but then DS confirmed during testing its much easier to work with in flat structure (Fivetran does not auto-flatten the same way Segment did) so they strongly preferred this version